### PR TITLE
Bugfix: create2 opcode should use deterministically derived evm contract address

### LINKED ISSUE #426: bug: the EVM address of a deployed contract is wrong
# Bug Report

**Current behavior:**

The EVM address of a contract account is computed as follows:

```
        // Generate EVM_contract address from the new cairo contract
        // TODO: Use RLP to compute proper EVM address, see https://www.evm.codes/#f0
        let (_, low) = split_felt(starknet_contract_address);
        local evm_contract_address = 0xAbdE100700000000000000000000000000000000 + low;
```

**Expected behavior:**

Depending on wether the call is a `CREATE` or a `CREATE2`, the address should be computed with

- CREATE (see https://www.evm.codes/#f0)
```
address = keccak256(rlp([sender_address,sender_nonce]))[12:]
```
- CREATE2 (see https://www.evm.codes/#f5)
```
address = keccak256(0xff + sender_address + salt + keccak256(initialisation_code))[12:]
```

**Related code:**

See `ContractAccount.deploy` function.
RLP utils are added in https://github.com/sayajin-labs/kakarot/pull/422

### LINKED ISSUE #427: bug: CREATE2 address does not depends on initialization bytecode and sender
# Bug Report

**Current behavior:**

The `CREATE2` opcode (see [CREATE2](https://www.evm.codes/#f5?fork=grayGlacier)) is meant to return a deterministic contract. address based on `deployer_address`, `salt` and `initialization_bytecode`.

Currently, the `exec_create2` does retrieve the `salt` and bytecode from the stack but the corresponding deployed starknet contract is deployed using only `salt`. Consequently, two different `CREATE2` with only the `salt` in common will result in the same starknet address.

This is due to the fact that `ContractAccount.deploy` does not use the bytecode nor the sender address to deploy the contract.

**Expected behavior:**

All the parameters should be used to compute the starknet address.

**Related code:**

A possible fix could be to update the `salt` passed to 

```
let (evm_contract_address, starknet_contract_address) = ContractAccount.deploy(salt);
```

to the hash of all the parameters `ctx.evm_contract_address, salt, bytecode`

**Other information:**

See also https://github.com/sayajin-labs/kakarot/issues/426

### PR DESCRIPTION
**Pull request by jobez**
Time spent on this PR: 1.75 days

## Pull request type

Please check the type of change your PR introduces:

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The EVM address of a contract account is computed as follows:

```
        // Generate EVM_contract address from the new cairo contract
        // TODO: Use RLP to compute proper EVM address, see https://www.evm.codes/#f0
        let (_, low) = split_felt(starknet_contract_address);
        local evm_contract_address = 0xAbdE100700000000000000000000000000000000 + low;
```

Resolves #427 

also resolves #426 wrt create2 opcode

## What is the new behavior?

Create2's address is defined via
```
address = keccak256(0xff + sender_address + salt + keccak256(initialisation_code))[12:]
```
-
-
-

## Other information

There is a current temporary split in deploy code, where create2 logic is handled separately from create. When both construct their address deterministically, the temporary refactor will be removed.

### CODE REVIEW

File: src/kakarot/instructions/system_operations.cairo [hunk 1]
```diff
@@ -72,16 +73,38 @@ namespace SystemOperations {
...
 
-        // create2 dynamic gas:
+        // create dynamic gas:
         // dynamic_gas = 6 * minimum_word_size + memory_expansion_cost + deployment_code_execution_cost + code_deposit_cost
         // -> ``memory_expansion_cost + deployment_code_execution_cost + code_deposit_cost`` is handled inside ``initialize_sub_context``
         let (minimum_word_size) = Helpers.minimum_word_count(size.low);
         let word_size_gas = 6 * minimum_word_size;
         let ctx = ExecutionContext.increment_gas_used(self=ctx, inc_value=word_size_gas);
 
+        let (evm_contract_address, starknet_contract_address) = ContractAccount.deploy_create(
```
Comment (jobez) [inline src/kakarot/instructions/system_operations.cairo:83]: create opcode temporarily still uses prior deployment conventions

File: src/kakarot/instructions/system_operations.cairo [hunk 2]
```diff
@@ -72,16 +73,38 @@ namespace SystemOperations {
...
+            _salt
+        );
+
+        let (bytecode: felt*) = alloc();
```
Comment (jobez) [inline src/kakarot/instructions/system_operations.cairo:87]: since create2's address construction depends on bytecode, i refactored `CreateHelper.initialize_sub_context` to take bytecode as argument

File: src/kakarot/instructions/system_operations.cairo [hunk 3]
```diff
@@ -548,6 +597,101 @@ namespace CallHelper {
...
+            let (local concat_args_bytes: felt*) = alloc();
+            // 0xff is by convention the marker involved in deterministic address creation for create2
+
+            let (concat_args_bytes_len) = Helpers.felt_to_bytes(0xff, 0, concat_args_bytes);
+
+            let (concat_args_bytes_len) = Helpers.felt_to_bytes(
+                sender_address, concat_args_bytes_len, concat_args_bytes + concat_args_bytes_len
+            );
+
+            let (salt_bytes_array_len, salt_bytes_array) = Helpers.uint256_to_bytes_array(salt);
```
Comment (jobez) [inline src/kakarot/instructions/system_operations.cairo:649]: Treating the salt as a uin256 ensures we get a padding of 32 bytes, which I believe is necessary.

see: https://github.com/ethereum/solidity/issues/12811

File: src/kakarot/instructions/system_operations.cairo [hunk 4]
```diff
@@ -548,6 +597,101 @@ namespace CallHelper {
...
+            memcpy(
+                dst=concat_args_bytes + concat_args_bytes_len,
+                src=salt_bytes_array,
+                len=salt_bytes_array_len,
+            );
+
+            let (
+                bytecode_hash_bytes_array_len, bytecode_hash_bytes_array
+            ) = Helpers.uint256_to_bytes_array(bytecode_hash_bigend);
+            memcpy(
```
Comment (jobez) [inline src/kakarot/instructions/system_operations.cairo:659]: I had the thought of writing a helper variant of `Helpers.uint256_to_bytes_array` that writes the bytes array to a destination array.

It might help make this code more readable, as otherwise you have to 'manually' track the length of `concat_args_bytes`

File: tests/integration/helpers/helpers.py [hunk 5]
```diff
@@ -80,7 +80,10 @@ def get_create2_address(
     """
     return to_checksum_address(
         keccak(
-            b"\xff" + decode_hex(sender_address) + salt + keccak(initialization_code)
+            b"\xff"
+            + to_bytes(hexstr=sender_address)
```
Comment (jobez) [inline tests/integration/helpers/helpers.py:84]: I need to investigate more why this change was necessary, but here is some output of my investigations.

I was testing locally before pulling main with

```
def create2_address(caller, bytecode, salt) -> bytes:
    code_hash = keccak(bytes(bytecode))
    data = b"\xff" + to_bytes(caller) + to_bytes(salt) + code_hash
    address = keccak(data)
    return address
```

The `get_create2_address` helper was giving me different results. I inspected it in a debugger session and saw that using `to_bytes` gave equivalent addresses

```
(Pdb) to_checksum_address(keccak(b"\xff" + to_bytes(hexstr=hex(evm_caller_address)) + padded_salt + keccak(bytecode))[-20:])
'0x78EA4300b4e2Ed68ac4BaE23059a825932755c72'
(Pdb) to_checksum_address(expected_create2_addr[-20:])
'0x78EA4300b4e2Ed68ac4BaE23059a825932755c72'
```

File: src/kakarot/instructions/system_operations.cairo [hunk 6]
```diff
@@ -548,6 +538,129 @@ namespace CallHelper {
...
+    // @param bytecode_len - The length of the initialization code.
+    // @param nonce - The nonce given to the create opcode.
+    // @return The pointer to the updated calling context.
+    func get_create_address{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(sender_address: felt, nonce: felt) -> (evm_contract_address: felt) {
+        // dummy impl to aide refactoring
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:555]: can you just add a `TODO` and a link an issue?

File: src/kakarot/instructions/system_operations.cairo [hunk 7]
```diff
@@ -548,6 +538,129 @@ namespace CallHelper {
...
+            // marker + caller_address + salt + bytecode_hash
+            let (local packed_bytes: felt*) = alloc();
+
+            // 0xff is by convention the marker involved in deterministic address creation for create2
+            let (packed_bytes_len) = Helpers.felt_to_bytes(0xff, 0, packed_bytes);
+
+            let (sender_address_high, sender_address_low) = split_felt(sender_address);
+            let (
+                sender_address_bytes_array_len, sender_address_bytes_array
+            ) = Helpers.uint256_to_bytes_array(
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:607]: isn't there a `felt_to_bytes_array`?

File: src/kakarot/instructions/system_operations.cairo [hunk 8]
```diff
@@ -548,6 +538,129 @@ namespace CallHelper {
...
+                Uint256(low=sender_address_low, high=sender_address_high)
+            );
+
+            // the address should be twenty bytes, so we skip the leading 12 elements
+            // and copy with a length of twenty
+            tempvar just_twenty_bytes_len = 20;
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:613]: do we really need this `tempvar`?
maybe somewhere in `constants` you can add a `const ADDRESS_BYTES_LEN = 20`?

File: src/kakarot/instructions/system_operations.cairo (hunk 4 above)
Comment (ClementWalter) [reply src/kakarot/instructions/system_operations.cairo:630]: yes, I agree it feels like we could have another helper that would factorize this

File: src/kakarot/instructions/system_operations.cairo [hunk 9]
```diff
@@ -548,6 +538,129 @@ namespace CallHelper {
...
+            let (create2_hash) = keccak_bigend(inputs=packed_bytes8, n_bytes=packed_bytes_len);
+
+            finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+        }
+        // this is the equivalent of a slice of the last 20 elements of a byte array
+        // i.e. in python address[-20:]
+        // This will keep the last 20 bytes of the byte array and set the rest of the bytes to 0 (here encoded as a uint256)
+        let mask = Helpers.to_uint256(0xffffffffffffffffffffffffffffffffffffffff);
+
+        let (create2_address_uint256) = uint256_and(a=create2_hash, b=mask);
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:659]: maybe better to handle directly `low` and `high`:

```
(_, r) = unsigned_div_rem(create2_hash.high, 256**4)
create2_address = create2_hash.low + r * 2**128
```

File: src/kakarot/instructions/system_operations.cairo [hunk 10]
```diff
@@ -556,31 +669,68 @@ namespace CreateHelper {
...
         let ctx = ExecutionContext.increment_gas_used(
             self=ctx, inc_value=gas_cost + SystemOperations.GAS_COST_CREATE
         );
 
+        // create2 context pops 4 off the stack, create pops 3
+        // so we use popped_len to derive the way we should handle
+        // the creation of evm addresses
+        local evm_contract_address;
+
+        if (popped_len != 4) {
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:695]: let's do this `if` at the very end to return in both branches and avoid all the `tempvar` (Lucas Levy advises me to do this one day)

File: src/kakarot/library.cairo [hunk 11]
```diff
@@ -191,9 +192,8 @@ namespace Kakarot {
     ) {
         alloc_locals;
         let (current_salt) = salt.read();
-        let (evm_contract_address, starknet_contract_address) = ContractAccount.deploy(
-            current_salt
-        );
+        let (evm_contract_address) = CreateHelper.get_create_address(0, current_salt);
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:195]: use `IAccount.get_eth_address(caller)` instead of `0`, (or add a TODO and an issue)

**Review by ClementWalter — CHANGES_REQUESTED** [review]

File: src/kakarot/instructions/system_operations.cairo (hunk 7 above)
Comment (jobez) [reply src/kakarot/instructions/system_operations.cairo:607]: I initially was using that, but ran into an overflow error that came up when the testrunner ran this test: https://github.com/sayajin-labs/kakarot/blob/main/tests/integration/solidity_contracts/PlainOpcodes/test_plain_opcodes.py#L349

You can see the runtime error output here: https://github.com/sayajin-labs/kakarot/actions/runs/3955455088/jobs/6773922685#step:9:477

My interpretation was that because the address as encoded into a felt had overflow when dividing by 256: https://github.com/sayajin-labs/kakarot/blob/734606cf99777cb040bb8926fac4c5ada89bdc3f/src/utils/utils.cairo#L622

Let me know if my interpretation was wrong here.

File: src/kakarot/library.cairo (hunk 11 above)
Comment (jobez) [reply src/kakarot/library.cairo:195]: Nice. This was a bit I was unsure on!

File: src/kakarot/instructions/system_operations.cairo [hunk 12]
```diff
@@ -548,6 +538,127 @@ namespace CallHelper {
...
+            let (create2_hash) = keccak_bigend(inputs=packed_bytes8, n_bytes=packed_bytes_len);
+
+            finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+        }
+        // this is the equivalent of a slice of the last 20 elements of a byte array
+        // i.e. in python address[-20:]
+        // This will keep the last 20 bytes of the byte array and set the rest of the bytes to 0 (here encoded as a uint256)
+        let mask = Helpers.to_uint256(0xffffffffffffffffffffffffffffffffffffffff);
+
+        // let (create2_address_uint256) = uint256_and(a=create2_hash, b=mask);
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:657]: should remove this?

File: src/kakarot/library.cairo [hunk 13]
```diff
@@ -191,9 +192,12 @@ namespace Kakarot {
     ) {
         alloc_locals;
         let (current_salt) = salt.read();
-        let (evm_contract_address, starknet_contract_address) = ContractAccount.deploy(
-            current_salt
+        let (current_address) = get_caller_address();
+        let (sender_eth_address) = IAccount.get_eth_address(current_address);
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:196]: 👍

File: tests/unit/src/kakarot/instructions/test_system_operations.py [hunk 14]
```diff
@@ -79,9 +82,34 @@ async def test_create(
     async def test_create2(
         self, system_operations, contract_account_class, account_registry
     ):
-        await system_operations.test__exec_create2__should_return_a_new_context_with_bytecode_from_memory_at_empty_address(
+        # we store a memory word in memory
+        # and have our bytecode as the memory read from an offset and size
+        # we take that word at an offset and size and use it as the bytecode to determine the expected create2 evm contract address
+        # Word is 0x 11 22 33 44 55 66 77 88 00 00 ... 00
+        # bytecode should be 0x 44 55 66 77
+        memory_word = 22774453838368691922685013100469420032
```
Comment (ClementWalter) [inline tests/unit/src/kakarot/instructions/test_system_operations.py:90]: fyi this could actually be write `memory_word = 0x11223344556677880000...` without quote (I think that this is copy/pasted from elsewhere)

**Review by ClementWalter — CHANGES_REQUESTED** [review]

**Review by ClementWalter — APPROVED** [review]

File: src/kakarot/instructions/environmental_information.cairo [hunk 15]
```diff
@@ -163,8 +163,9 @@ namespace EnvironmentalInformation {
         alloc_locals;
         // Get caller address.
         let (current_address) = get_caller_address();
-        let caller_address = Helpers.to_uint256(current_address);
-        let stack: model.Stack* = Stack.push(self=ctx.stack, element=caller_address);
+        let (evm_address) = IAccount.get_eth_address(current_address);
```
Comment (jobez) [inline src/kakarot/instructions/environmental_information.cairo:166]: these changes resolve #438 , right?

Comment (ClementWalter) [reply src/kakarot/instructions/environmental_information.cairo:166]: indeed

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]
